ga4gh/utils/seqfetcher.py

```diff
@@ -0,0 +1,207 @@
+     ('ENST00000288602', u'CGCCTCCCTTCCCCCTCCCCGCCCG'),
+     ('ENSP00000288602', u'MAALSGGGGGGAEPGQALFNGDMEP')]
+
+
+    RuntimeError is thrown in the case of errors::
+
+    >>> fetch_seq('NM_9.9')
+    Traceback (most recent call last):
+       ...
+    RuntimeError: No sequence available for NM_9.9
+
+    >>> fetch_seq('QQ01234')
+    Traceback (most recent call last):
+       ...
+    RuntimeError: No sequence fetcher for QQ01234
+
+    """
+
+    ac_dispatch = [
+        {'re': re.compile('^(?:AC|N[CGMPRTW])_|^[A-L]\w\d|^U\d'),
+         'fetcher': _fetch_seq_ncbi},
+        {'re': re.compile('^ENS[TP]\d+'),
+         'fetcher': _fetch_seq_ensembl},
+        ]
+
+    eligible_fetchers = [dr['fetcher']
+                         for dr in ac_dispatch
+                         if dr['re'].match(ac)]
+
+    if len(eligible_fetchers) == 0:
+        raise RuntimeError("No sequence fetcher for {ac}".format(ac=ac))
+
+    elif len(eligible_fetchers) == 1:
+        fetcher = eligible_fetchers[0]
+
+    elif len(eligible_fetchers) >= 1:
+        logger.debug("Multiple sequence fetchers found for "
+                     "{ac}; using first".format(ac=ac))
+        fetcher = eligible_fetchers[0]
+
+    logger.debug("fetching {ac} with {f}".format(
+        ac=ac, f=fetcher))
+
+    try:
+        return fetcher(ac, start_i, end_i)
+    except requests.HTTPError:
+        raise RuntimeError("No sequence available for {ac}".format(ac=ac))
+
+
+# ###########################################################################
+# Internal functions
+
+def _fetch_seq_ensembl(ac, start_i=None, end_i=None):
+    """Fetch the specified sequence slice from Ensembl using the public
+    REST interface.
+
+    An interbase interval may be optionally provided with start_i and
+    end_i. However, the Ensembl REST interface does not currently
+    accept intervals, so the entire sequence is returned and sliced
+    locally.
+
+    >>> len(_fetch_seq_ensembl('ENSP00000288602'))
+    766
+
+    >>> _fetch_seq_ensembl('ENSP00000288602',0,10)
+    u'MAALSGGGGG'
+
+    >>> _fetch_seq_ensembl('ENSP00000288602')[0:10]
+    u'MAALSGGGGG'
+
+    >>> ac = 'ENSP00000288602'
+    >>> _fetch_seq_ensembl(ac ,0, 10) == _fetch_seq_ensembl(ac)[0:10]
+    True
+
+    """
+
+    url_fmt = "http://rest.ensembl.org/sequence/id/{ac}"
+    url = url_fmt.format(ac=ac)
+    r = requests.get(url, headers={"Content-Type": "application/json"})
+    r.raise_for_status()
+    seq = r.json()['seq']
+    return seq if (start_i is None or end_i is None) else seq[start_i:end_i]
+
+
+def _fetch_seq_ncbi(ac, start_i=None, end_i=None):
+    """Fetch sequences from NCBI using the eutils interface.
+
+    An interbase interval may be optionally provided with start_i and
+    end_i. NCBI eutils will return just the requested subsequence,
+    which might greatly reduce payload sizes (especially with
+    chromosome-scale sequences).
+
+    >>> len(_fetch_seq_ncbi('NP_056374.2'))
```

Comment:
I tend to agree with @dcolligan here. While the doctest comments are useful, are they likely to be fully maintained in the future? I guess we'd need to actually run the doctests as part of CI to ensure this. However, we'll then probably have random CI failures when NCBI is overloaded and so on... It's a tricky one!